File: dispatch/p1_cpp_dispatch/s2_struct.cpp
struct Square {
    int side;
};

struct Triangle {
    int side;
    int height;
};

int area_square(Square s) {
    return s.side * s.side;
}

int area_triangle(Triangle t) {
    return (t.side * t.height)/2;
}

int main() {
    auto a1 = area_square({5});
    auto a2 = area_triangle({10, 20});
    return a1 + a2;
}

File: dispatch/p2_rust_dispatch/s2_struct.rs
struct Square {
    side: i32
}

impl Square {
    pub fn new(side: i32) -> Self {
        Self { side }
    }
    pub fn area(&self) -> i32 {
        self.side * self.side
    }
}

struct Triangle {
    side: i32,
    height: i32
}

impl Triangle {
    pub fn new(side: i32, height: i32) -> Self {
        Self { side, height }
    }
    pub fn area(&self) -> i32 {
        (self.side * self.height)/2
    }
}

pub fn main() -> i32 {
    let v1 = Square::new(5);
    let v2 = Triangle::new(10, 20);
    v1.area() + v2.area()
}
